### PR TITLE
fix(NODE-4532): round trip double values consistently

### PR DESCRIPTION
**Pull request by nbbeeken**
## Description

**What changed?**

Instead of relying on casting to convert our double I changed the code to use math.h or cmath's floor function.

### CODE REVIEW

File: src/bson.cc [hunk 1]
```diff
@@ -337,11 +337,14 @@ void BSONSerializer<T>::SerializeValue(void *typeLocation,
   // Process all the values
   if (value->IsNumber()) {
     double doubleValue = NanTo<double>(value);
-    int intValue = (int)doubleValue;
-    if (intValue == doubleValue) {
+    double flooredValue = floor(doubleValue);
+    if (flooredValue == doubleValue && flooredValue <= INT32_MAX && flooredValue >= INT32_MIN) {
```
Comment (addaleax) [inline src/bson.cc]: The previous code didn't handle this either, but what's the right behavior for -0? With the JS version we get the same behavior, but I'm not sure if it's the "correct" one:

```js
> bson.deserialize(bson.serialize({ a: -0 }))
{ a: 0 }
> bson.deserialize(bson.serialize({ a: bson.Double(-0) }))
{ a: -0 }
```

**Review by addaleax — APPROVED** [review]

File: src/bson.cc (hunk 1 above)
Comment (nbbeeken) [reply src/bson.cc]: Very good point, thanks for bringing it up, I propose we align with js-bson and serialize to int32 for consistency, its least surprising I suppose. The Double class as a wrapper is still an avenue that can be used to preserve this edge case, until a future where we consider if this is the correct choice.

Comment (addaleax) [reply src/bson.cc]: To be clear, this was also what this code was doing before, but I do like that we're being explicit about it now 👍

File: src/bson.cc [hunk 2]
```diff
@@ -337,11 +337,21 @@ void BSONSerializer<T>::SerializeValue(void *typeLocation,
   // Process all the values
   if (value->IsNumber()) {
     double doubleValue = NanTo<double>(value);
-    int intValue = (int)doubleValue;
-    if (intValue == doubleValue) {
+    double flooredValue = floor(doubleValue);
```
Comment (addaleax) [inline src/bson.cc]: teeny tiny suggestion, feel free to ignore, just makes this a bit more consistent with the line below it

```suggestion
    double flooredValue = std::floor(doubleValue);
```

**Review by addaleax — APPROVED** [review]

File: test/node/number_test.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+const nanPayloadBuffer = Buffer.from("120000000000F87F", "hex");
+const nanPayloadDV = new DataView(
+  nanPayloadBuffer.buffer,
+  nanPayloadBuffer.byteOffset,
+  nanPayloadBuffer.byteLength
+);
+const nanPayloadDouble = nanPayloadDV.getFloat64(0, true);
+
+describe("serializing javascript numbers", () => {
+  it("should serialize a number that exceeds int32.max as a double", () => {
```
Comment (durran) [inline test/node/number_test.js]: Can we change the description language in these tests not to use "should" as in our testing guidelines?

**Review by durran — CHANGES_REQUESTED** [review]

File: src/bson.cc (hunk 2 above)
Comment (nbbeeken) [reply src/bson.cc]: Worth doing, fixed 🙂

File: test/node/number_test.js (hunk 3 above)
Comment (nbbeeken) [reply test/node/number_test.js]: updated!

**Review by durran — APPROVED** [review]

**Pull request merged**